### ISSUE TITLE
"Unable to get property 'nodeType' of undefined or null reference" when deleting last remaining character of paragraph

### ISSUE THREAD

**Issue by andreash82** [ProseMirror/prosemirror#930]
### Issue details

Target browser: IE11, with prosemirror-view 1.9.6
Having any paragraph (or line after linebreak) with a single character, deleting this character (without selecting it first) results in error "SCRIPT5007:  Unable to get property 'nodeType' of undefined or null reference".

I could track it down to different inputs & results in DOMObserver.flush.
Deleting a character results in following "steps":
- IE11
  1. flush called with 1 MutationRecord (childList) => from > 0, to > 0, newSel = false
  2. flush called without any MutationRecord, also none from charDataQueue
    - currentSelection = Textnode
    - root selection = Paragraph
    - from & to  = -1, newSel = true
    - -> **error via selectionToDOM -> setSelection -> isEquivalentPosition -> scanFor -> nodeSize(node)**
- for comparison in Chrome:
  1. flush called with an array of 3 MutationRecords (characterData, childList, childList) => from > 0, to > 0, newSel = true
  2. flush called without any MutationRecord
    - currentSelection = Paragraph
    - root selection = Paragraph
    - from & to = -1, newSel = false -> no error
  - ... further flush calls due to other stuff.

When selecting (marking) & deleting the letter -> no flush call at all, no error.

Yes, I know IE11 is crap, but some people still have it as requirement. 

### Steps to reproduce

1. Clone prosemirror/website (since used deps on prosemirror.net aren't updated yet)
2. Update package.json to current prosemirror versions
`"prosemirror-collab": "^1.0.0",
    "prosemirror-commands": "^1.0.8",
    "prosemirror-example-setup": "^1.0.1",
    "prosemirror-history": "^1.0.4",
    "prosemirror-inputrules": "^1.0.4",
    "prosemirror-keymap": "^1.0.1",
    "prosemirror-gapcursor": "^1.0.3",
    "prosemirror-markdown": "^1.0.0",
    "prosemirror-menu": "^1.0.0",
    "prosemirror-model": "^1.7.0",
    "prosemirror-schema-basic": "^1.0.1",
    "prosemirror-schema-list": "^1.0.3",
    "prosemirror-state": "^1.2.3",
    "prosemirror-transform": "^1.1.3",
    "prosemirror-view": "^1.9.6"
`
3. make and start app
4. in IE11 open this app, open console

Test with paragraph
- open/refresh page
- start a new paragraph in editor
- type one character
- delete this again (via backspace or delete)
- => console shows error "SCRIPT5007:  Unable to get property 'nodeType' of undefined or null reference"
- after that, it also logs this error when clicking into the editor's empty space

Test with linebreak
- open/refresh page
- start a new paragraph
- type some characters 
- press shift+enter to insert hard break
- type one character
- delete this letter
- => console shows error "SCRIPT5007:  Unable to get property 'nodeType' of undefined or null reference"
- type a letter => console shows error "invalid argument", with optional weird behaviour after that 

### ProseMirror version

prosemirror-view 1.9.6 (updated after to #927 )

### Affected platforms

- [ ] Chrome
- [ ] Firefox
- [x] Internet Explorer 11
- [ ] Other

**Comment by marijnh** [comment]
Does changing line 37 in `scanFor` to

    if (!parent || parent.nodeType != 1 || hasBlockDesc(node) || atomElements.test(node.nodeName) || node.contentEditable == "false")

help? It's highly suspicious that there's no mutation events when you deleted a character, but possibly those will be fired afterwards.

Can you see who called `flush`?

**Comment by andreash82** [comment]
This change doesn't help but should probably remain there for safety reasons ;). 

In line 42 `node.childNodes[...] `returns null cause of `off=1, dir=-1, node=paragraph (already empty)`which leads to the error.
Inserting  `if (!node) { return true; }` before line 43 or modifying the else-block to
```
if (node.childNodes.length == 0) {
  off = 0;
} else {
  node = node.childNodes[off + (dir < 0 ? -1 : 0)];
  off = dir < 0 ? nodeSize(node) : 0;
}
```
works, with the difference to other browsers or empty blocks, that this paragraph remains completely emtpy (`<p></p>`) instead of the typical `<p><br/></p>` structure (which also seems to be a reason for further errors, since this is not allowed).
Inserting `return false` instead results in an Index error in viewdesc.js:356 with headDOM.node being the paragraph and its offset = 1

In the erroneous case flush is called by onSelectionChange. This handler was "reconnected" after flush -> EditorView.updateState -> updateStateInner in selectionToDOM (visible as "asynchronous call" in IE Devtools Stack.)

**Comment by marijnh** [comment]
> This change doesn't help but should probably remain there for safety reasons ;).

Well, the function currently assumes that it's called with nodes that are in the document (which should hold unless the DOM selection ends up pointing at orphan nodes... which I wouldn't put past IE, but until we have evidence that that happens I don't think we need to account for it).

> In line 42 node.childNodes[...] returns null cause of off=1, dir=-1, node=paragraph (already empty)which leads to the error.

That's bad, that `off==1`, since that's not a valid offset for an empty parent node. I guess the problem is that the DOM _did_ change, the browser just didn't notify us yet (even though the code asked through `takeRecords()`), and the code is proceeding with the assumption that the view desc tree accurately reflects the current DOM structure when in fact some nodes in there have been removed.

If you somehow suppress this error (possibly by just not calling `flush` on selection change), does a mutation come in through the mutation observer right after the selection event? If so, maybe adding a slight delay to the handling of selectionchange in IE would be a reasonable workaround.

**Comment by andreash82** [comment]
It is the only "flush" call up to any other changes.
I tried to delay addEventListener('selectionchange') in connectSelection, which helped to fix this path to the error, but resulted in later errors (during cursor movement, editorfocus etc).

All those errors where related to the default "delete empty textblock" behaviour of IE11 with leaving an empty `<p></p>`, while deletion with selected text worked fine.
So I looked deeper into the code and found `stopNativeHorizontalDelete` in capturekeys.js.

With change of line 193 to
```
if (nextNode && (!nextNode.isText || (nextNode.nodeSize == 1 && browser.ie && browser.ie_version <= 11))) {
```
everything worked fine (I just tested it on dist/index.js without a full prosemirror dev env, but that snippet and line matches the source file in my node_modules directory).

In conclusion, we just have to force a replacement with an empty Slice in IE.

**Comment by marijnh** [comment]
> In conclusion, we just have to force a replacement with an empty Slice in IE.

Right, but I'd rather not depend on specific keys being pressed in handling this. A change to the DOM happening but not firing any observer events is _scary_.

I guess I'll have to reinstall my VM and dig into this. Might be a while until I get to that though.

**Comment by andreash82** [comment]
If it helps, I've just setup a small jsbin to track all relevant events without code based influences/changes: http://jsbin.com/maqoqofayi/edit?js,console,output

Observations in IE:
- start new paragraph by typing enter: triple childList records
- enter 1 character: childList record (created Textnode)
- delete that character: no mutationrecord or character event at all, but correct `<p><br></p>`
- characterData is first triggered with second character, but then also for all deletions with `<p><br/></p>` as final result

**Comment by marijnh** [comment]
I _finally_ got around to debugging this, and https://github.com/ProseMirror/prosemirror-view/commit/a607dc717990a4f2fa909c6c98a99e8360625474 should fix this.

**Comment by andreash82** [comment]
This fixes it. Thanks!

Only one small thing is remaining: The last error in the test with the linebreak:

- open/refresh page
- start a new paragraph
- type some characters
- press shift+enter to insert hard break
- type one character
- delete this letter
- **type a letter => console shows error "invalid argument"**

This refers to `renderDescs -> parentDOM.insertBefore(childDOM, dom)` where dom is null and childDOM an empty text node, having as `descs` argument an array of Textnode (with text), BR element and the empty textnode

Probably another issue?

**Comment by marijnh** [comment]
That was another issue -- though maybe the first was a symptom of the second. Attached patch seems to help.

Please open a new issue if you run into additional errors.

**Comment by tommoor** [comment]
I seem to really reproducibly have the same issue when using a NodeView that contains an iframe. I get `TypeError: Cannot read property 'nodeType' of null` just by making a selection that contains the NodeView.

A simpler `NodeView` (eg an `img` tag) doesn't exhibit the issue. Does that make any sense based on your knowledge of how this method works?

**Comment by GuySerfaty** [comment]
We faced something similar
```
Uncaught TypeError: Cannot read properties of null (reading 'nodeType')
```
With a nodeView, when deleting the last remaining character of the contentDOM.

In the end, we found out that accidentally we set the contentDOM with display: inline (we used <span>)
I believe the Browser put the cursor inside an element that is not connected to the dom anymore - the parent is null.

(https://github.com/ProseMirror/prosemirror/issues/741#issuecomment-352363393 this one helped me a lot)
